eclipse-theia/theia
# filesystem: upload via HTTP

Refactor the code to upload files via HTTP rather than WS. This
simplifies the code by removing the ad-hoc WS upload protocol and should
make it easier to configure proxies as it will be using good ol' HTTP.

Note 1: The file upload is done over HTTP POST `multipart/form-data`.

Note 2: I had to use the older XMLHttpRequest (XHR/AJAX) API in order to
monitor the upload progress. The Fetch API is not there yet.

Add new preference `files.maxConcurrentUploads` that defaults to 1 as it
seems to the be fastest on my test environment. Can be user-customized.

Add new class `AsyncQueue` that runs a fixed amount of async tasks
concurrently. We cannot use `p-queue` in the browser, hence this minimal
implementation.

This PR goes hand-in-hand with https://github.com/eclipse-theia/theia/pull/9731

#### How to test

- Upload files by dnd over the file explorer, many should work.
- Upload files using the contextual command (right click in the file navigator), many should work.
- Progress notification should be somewhat accurate in both cases.

#### Review checklist


#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)
